repository.py

```diff
@@ -364,6 +364,22 @@ def stage_charm(
         except OSError:
             raise RepositoryError(f"Failed to write file `{charm.build_path / CHARMCRAFT_FILE}`")
 
+        # Create a version file and pack it into the charm. This is dynamically generated to ensure
+        # that the git revision of the charm is always recorded in this version file.
+        git_hash = subprocess.check_output(
+            [
+                "git",
+                "-C",
+                str(charm.path),
+                "describe",
+                "--always",
+                "--dirty",
+            ],
+            stderr=subprocess.STDOUT
+        ).strip().decode("utf-8")
+        version_file = Path(charm.build_path / "version")
+        version_file.write_text(git_hash)
```

Comment:
I'd suggest using `BuildTool` like we do for `uv` and `charmcraft`, since it gives nice logs of what's happening + handles decoding for us.